base_layer/core/src/base_node/sync/header_sync/validator.rs

```diff
@@ -105,23 +105,21 @@ impl<B: BlockchainBackend + 'static> BlockHeaderSyncValidator<B> {
             timestamps,
             target_difficulties,
             previous_accum,
+            // One large allocation is usually better even if it is not always used.
+            valid_headers: Vec::with_capacity(1000),
```

Comment:
I disagree here. I think this is too large. 
Headers can be quite large I think we should start with a more reasonable number that is just above the median. Perhaps around 200 for now, main net should be way less

Reply:
Yeah this wasn't particularly thought out (as in I didn't try to size up the block header struct to get an exact memory cost - I will now and report) although it is true that many small allocations is significantly slower than one large one. Remember this will allocate the "shallow" size of the struct (i.e. a vec on the heap is not part of what is allocated in the `valid_headers` vec) 

Because there is really (easy/reliable) no way to know for sure what is the optimal value - for initial sync it should be 1000 (the interval at which headers are "flushed") - if we're doing a large reorg it could grow to 1000s before being committed - if a node is slightly behind this is wasteful. 

I'll report on the exact cost of preallocation.

Reply:
The "shallow size" of the block header struct is 208 bytes (`std::mem::size_of::<BlockHeader>()`) So 208kb is allocated as a single block (this excludes any Vecs which are allocated separately)

Reply:
I don't think that it's worth putting engineering effort into optimising a potentially wasteful 208kb allocation while in sync mode (the base node only syncs every, say, few minutes at the very worst). We may want to add a temporary an LMDB database for this purpose, eliminating the issue entirely but I didn't want to include changes to the blockchain backend when an in-memory variable likely suffices.

Reply:
Agreed.